### ISSUE TITLE
Connection reset by peer

### ISSUE THREAD

**Issue by wilfriedroset** [zalando/patroni#857]
Hi,

We have encountered today a similar bug to #393. Patroni API stop responding for approximately 10min, hence no probes and therefore no more queries. All traffic flow restart just after the time of the following traceback:

```
Nov 08 15:50:17 patroni-node-1 patroni[12390]: Traceback (most recent call last):
Nov 08 15:50:17 patroni-node-1 patroni[12390]:   File "/usr/lib/python3.5/socketserver.py", line 625, in process_request_thread
Nov 08 15:50:17 patroni-node-1 patroni[12390]:     self.finish_request(request, client_address)
Nov 08 15:50:17 patroni-node-1 patroni[12390]:   File "/usr/lib/python3.5/socketserver.py", line 354, in finish_request
Nov 08 15:50:17 patroni-node-1 patroni[12390]:     self.RequestHandlerClass(request, client_address, self)
Nov 08 15:50:17 patroni-node-1 patroni[12390]:   File "/usr/lib/python3.5/socketserver.py", line 681, in __init__
Nov 08 15:50:17 patroni-node-1 patroni[12390]:     self.handle()
Nov 08 15:50:17 patroni-node-1 patroni[12390]:   File "/usr/lib/python3.5/http/server.py", line 422, in handle
Nov 08 15:50:17 patroni-node-1 patroni[12390]:     self.handle_one_request()
Nov 08 15:50:17 patroni-node-1 patroni[12390]:   File "/usr/lib/python3.5/http/server.py", line 400, in handle_one_request
Nov 08 15:50:17 patroni-node-1 patroni[12390]:     if not self.parse_request():
Nov 08 15:50:17 patroni-node-1 patroni[12390]:   File "/usr/lib/python3/dist-packages/patroni/api.py", line 389, in parse_request
Nov 08 15:50:17 patroni-node-1 patroni[12390]:     ret = BaseHTTPRequestHandler.parse_request(self)
Nov 08 15:50:17 patroni-node-1 patroni[12390]:   File "/usr/lib/python3.5/http/server.py", line 327, in parse_request
Nov 08 15:50:17 patroni-node-1 patroni[12390]:     "Bad request syntax (%r)" % requestline)
Nov 08 15:50:17 patroni-node-1 patroni[12390]:   File "/usr/lib/python3.5/http/server.py", line 477, in send_error
Nov 08 15:50:17 patroni-node-1 patroni[12390]:     self.wfile.write(body)
Nov 08 15:50:17 patroni-node-1 patroni[12390]:   File "/usr/lib/python3.5/socket.py", line 594, in write
Nov 08 15:50:17 patroni-node-1 patroni[12390]:     return self._sock.send(b)
Nov 08 15:50:17 patroni-node-1 patroni[12390]:   File "/usr/lib/python3.5/ssl.py", line 871, in send
Nov 08 15:50:17 patroni-node-1 patroni[12390]:     return socket.send(self, data, flags)
Nov 08 15:50:17 patroni-node-1 patroni[12390]: ConnectionResetError: [Errno 104] Connection reset by peer
```

Our probes are using already OPTIONS method
```
global
    log 127.0.0.1 local0
    stats socket /run/haproxy/admin.sock mode 660 level admin
    stats timeout 30s
    chroot /var/lib/haproxy
    server-state-base /var/lib/haproxy/state
    user  haproxy
    group haproxy

defaults
    log global
    load-server-state-from-file local
    mode tcp
    # Don't enable "%fi" or you will end with D process
    log-format "[%T] client=%ci:%cp frontend=%f backend=%b server=%si:%sp"
    retries 6
    timeout client 1d
    timeout connect 3s
    timeout server 1d
    timeout check 5s
    timeout tunnel 1d

listen stats
    bind localhost:1936
    mode http
    stats enable
    stats hide-version
    stats realm Haproxy\ Statistics
    stats uri /

listen read-write
    bind *:3000
    option httpchk OPTIONS /master
    maxconn 10000
    default-server fall 5 inter 1000 rise 5 downinter 1000 on-marked-down shutdown-sessions weight 10 maxconn 1000
    server patroni-backup-1 x.x.x.x:5432 check port 8008 check-ssl verify none
    server patroni-node-1 x.x.x.x:5432 check port 8008 check-ssl verify none
    server patroni-node-2 x.x.x.x:5432 check port 8008 check-ssl verify none


listen read-only
    bind *:4000
    option httpchk OPTIONS /replica
    maxconn 10000
    default-server fall 5 inter 1000 rise 5 downinter 1000 on-marked-down shutdown-sessions weight 10 maxconn 1000
    server read-write 127.0.0.1:3000 backup
    server patroni-backup-1 x.x.x.x:5432 check port 8008 check-ssl verify none
    server patroni-node-1 x.x.x.x:5432 check port 8008 check-ssl verify none
    server patroni-node-2 x.x.x.x:5432 check port 8008 check-ssl verify none
```

Unfortunately patroni was not in debug mode so there is no useful information in the log

Running on:
* debian 9.5
* postgresql 9.6
* patroni 1.5.0 (apt.postgresql.org repository)
* confd 0.14.0-2
* python3-kazoo 2.2.1
* zookeeper 3.4.5+dfsg-2+deb8u2

**Comment by CyberDem0n** [comment]
Can it be that postgres was very slow for some reason?
Do you monitor your servers?

**Comment by wilfriedroset** [comment]
Tl;dr: The impacted cluster was not slow or under heavy load

All of our patroni clusters are monitored and we also collect metrics using (telegraf)[https://github.com/influxdata/telegraf/tree/master/plugins/inputs].
In our graphs we can see a significant dent (drop to 0) for tup_* and xact_*. Which leads directly to a dent in term of load average and cpu usage. There is no sign of postgresql nor the server being slow/under heavy load.

We use several probes for generic system and (postgresql)[https://bucardo.org/check_postgres/check_postgres.pl.html] (sequence, txn-wraparound, backend, connections, replication and replication-slots) we have  a probe which check that patroni API respond.

```
API_HOST=localhost
API_PORT=8008
API_PROTO=https
API_PATH=
CERT_PATH=/etc/patroni/cert.pem
CONNECT_TIMEOUT=5

if ! curl -s --connect-timeout "${CONNECT_TIMEOUT}" --cacert "${CERT_PATH}" "${API_PROTO}://${API_HOST}:${API_PORT}/${API_PATH}" > /dev/null
then
    exit ${OCO_CRIT}
else
    exit ${OCO_OK}
fi
```

This homemade as be added a few months ago to monitor this type of API freeze.

**Comment by wilfriedroset** [comment]
Hi.

New occurrence of this problem today (usual workload, no new app, no new queries). This time we were faster to react. We have executed the following operation:
* patronictl pause
* systemctl restart patroni
* patroni resume

Everything start working fine again just after the restart of patroni.

**Comment by wilfriedroset** [comment]
Hi.

New occurrence of this problem today. We have recover using the same procedure (pause, restart, resume).

This time we were running patroni in debug.  We don't see anything unusual except from:

- INFO: Lock owner: patroni-48-node-[1-2]; I am patroni-48-node-[1-2]
- INFO: no action.
- DEBUG: Sending request
- DEBUG: Received response

Why do we say that we don't have anything unusual? The outage was at 3:15 PM. Using journalctl we have extract a subset of log

```
patroni-48-node-1 ~# journalctl -u patroni | grep 'Nov 19 15:' > patroni_freeze_20181119
```

And we proceed by filtering the resulting file.

Here is an example of Sending request/Received response lines

```
Nov 19 15:00:49 patroni-48-node-1 patroni[11424]: 2018-11-19 15:00:49,732 DEBUG: Sending request(xid=55709): SetData(path='/patroni/cluster48/optime/leader', data=b'6893385174560', version=-1)
Nov 19 15:00:49 patroni-48-node-1 patroni[11424]: 2018-11-19 15:00:49,734 DEBUG: Received response(xid=55709): ZnodeStat(czxid=17198457179, mzxid=17224239511, ctime=1535978271044, mtime=1542636049735, version=658873, cversion=0, aversion=0, ephemeralOwner=0, dataLength=13, numChildren=0, pzxid=17198457179)
--8<--
Nov 19 15:01:09 patroni-48-node-1 patroni[11424]: 2018-11-19 15:01:09,734 DEBUG: Sending request(xid=55714): SetData(path='/patroni/cluster48/members/patroni-48-node-1', data=b'{"api_url":"https://x.x.x.x:8008/patroni","conn_url":"postgres://x.x.x.x:5432/postgres","xlog_location":6893463973832,"state":"running","timeline":1,"role":"master"}', version=-1)
Nov 19 15:01:09 patroni-48-node-1 patroni[11424]: 2018-11-19 15:01:09,737 DEBUG: Received response(xid=55714): ZnodeStat(czxid=17223154622, mzxid=17224239591, ctime=1542358192025, mtime=1542636069738, version=27797, cversion=0, aversion=0, ephemeralOwner=99873831030490191, dataLength=173, numChildren=0, pzxid=17223154622)
```

The remaining lines are our probres (specially the one check the validity of the ssl cert) and the logs of patroni pause/restart/resume.

```
Nov 19 15:07:01 patroni-48-node-1 patroni[11424]: 2018-11-19 15:07:01,438 DEBUG: API thread: 127.0.0.1 - - [19/Nov/2018 15:07:01] "GET / HTTP/1.1" 200 -
--8<--
Nov 19 15:24:58 patroni-48-node-1 patroni[11424]: 2018-11-19 15:24:58,257 DEBUG: Received EVENT: Watch(type=3, state=3, path='/patroni/cluster48/config')
Nov 19 15:24:58 patroni-48-node-1 patroni[11424]: 2018-11-19 15:24:58,270 INFO: PAUSE: no action.  i am the leader with the lock
Nov 19 15:24:58 patroni-48-node-1 patroni[11424]: 2018-11-19 15:24:58,273 INFO: Changed DateStyle from ISO, MDY to iso, mdy
Nov 19 15:24:58 patroni-48-node-1 patroni[11424]: 2018-11-19 15:24:58,273 INFO: PostgreSQL configuration items changed, reloading configuration.
Nov 19 15:24:58 patroni-48-node-1 patroni[11424]: server signaled
Nov 19 15:25:08 patroni-48-node-1 patroni[11424]: 2018-11-19 15:25:08,267 INFO: PAUSE: no action.  i am the leader with the lock
Nov 19 15:25:18 patroni-48-node-1 patroni[11424]: 2018-11-19 15:25:18,264 INFO: PAUSE: no action.  i am the leader with the lock
Nov 19 15:25:18 patroni-48-node-1 systemd[1]: Stopping Runners to orchestrate a high-availability PostgreSQL...
Nov 19 15:25:26 patroni-48-node-1 patroni[11424]: 2018-11-19 15:25:26,948 DEBUG: Received EVENT: Watch(type=4, state=3, path='/patroni/cluster48/members')
Nov 19 15:25:48 patroni-48-node-1 systemd[1]: patroni.service: State 'stop-sigterm' timed out. Killing.
Nov 19 15:25:48 patroni-48-node-1 systemd[1]: patroni.service: Killing process 11424 (patroni) with signal SIGKILL.
Nov 19 15:25:48 patroni-48-node-1 systemd[1]: patroni.service: Main process exited, code=killed, status=9/KILL
Nov 19 15:25:48 patroni-48-node-1 systemd[1]: Stopped Runners to orchestrate a high-availability PostgreSQL.
Nov 19 15:25:48 patroni-48-node-1 systemd[1]: patroni.service: Unit entered failed state.
Nov 19 15:25:48 patroni-48-node-1 systemd[1]: patroni.service: Failed with result 'timeout'.
Nov 19 15:25:48 patroni-48-node-1 systemd[1]: Started Runners to orchestrate a high-availability PostgreSQL.
Nov 19 15:25:48 patroni-48-node-1 patroni[40256]: 2018-11-19 15:25:48,842 INFO: Failed to import patroni.dcs.consul
Nov 19 15:25:48 patroni-48-node-1 patroni[40256]: 2018-11-19 15:25:48,843 INFO: Failed to import patroni.dcs.etcd
Nov 19 15:25:48 patroni-48-node-1 patroni[40256]: 2018-11-19 15:25:48,860 INFO: Failed to import patroni.dcs.kubernetes
Nov 19 15:25:48 patroni-48-node-1 patroni[40256]: 2018-11-19 15:25:48,861 INFO: Connecting to x.x.x.x:xxxx
Nov 19 15:25:48 patroni-48-node-1 patroni[40256]: 2018-11-19 15:25:48,864 INFO: Zookeeper connection established, state: CONNECTED
Nov 19 15:25:48 patroni-48-node-1 patroni[40256]: 2018-11-19 15:25:48,877 INFO: I am leader but not owner of the session. Removing leader node
Nov 19 15:25:48 patroni-48-node-1 patroni[40256]: 2018-11-19 15:25:48,880 DEBUG: Received EVENT: Watch(type=4, state=3, path='/patroni/cluster48')
Nov 19 15:25:48 patroni-48-node-1 patroni[40256]: 2018-11-19 15:25:48,882 INFO: establishing a new patroni connection to the postgres cluster
Nov 19 15:25:48 patroni-48-node-1 patroni[40256]: server signaled
Nov 19 15:25:48 patroni-48-node-1 patroni[40256]: 2018-11-19 15:25:48,909 DEBUG: Received EVENT: Watch(type=4, state=3, path='/patroni/cluster48')
Nov 19 15:25:48 patroni-48-node-1 patroni[40256]: 2018-11-19 15:25:48,917 DEBUG: Received EVENT: Watch(type=4, state=3, path='/patroni/cluster48/members')
```

**Comment by jouir** [comment]
We use [confd](https://github.com/kelseyhightower/confd) to discover topology from zookeeper, update haproxy configuration and reload service when there's a checksum difference. We used a comment at the top of file to write current date. Date always change so haproxy service was reloaded very often (few minutes apart). We've removed this useless comment to release haproxy pressure.

haproxy is installed on every node of the cluster and we use keepalived/vrrp to dynamically move a vip on a valid node.

This morning, one patroni API on "node3" (anonymized name) has failed at approximately 05:01 but automatically recovered at approximately 05:19. Here is what we had in debug messages in between:
```
Jan 08 05:01:23 <node3> patroni[3725]: 2019-01-08 05:01:23,329 DEBUG: API thread: <node1> - - [08/Jan/2019 05:01:23] code 400, message Bad request syntax ("\x16\x03\x01\x02\x00\x01\x00\x01ü\x03\x03e&\x8eü\x89*\x1c¢\x0bv\\ÔZj\x89´¢¿¼mlf±\\s#æ2&9!1 ¥ÉMr\x97´þ&V½ÈvÄ®pº\x90fÜÿC\x03-ê\x9d\x0fOÅÑ 4Ò\x008À,À0\x00\x9fÌ©Ì¨ÌªÀ+À/\x00\x9eÀ$À(\x00kÀ#À'\x00gÀ")
Jan 08 05:02:01 <node3> patroni[3725]: 2019-01-08 05:02:01,132 DEBUG: API thread: <node2> - - [08/Jan/2019 05:02:01] code 400, message Bad request version ("¡¦r\x99\x91«}\x7f\x95ìÔùØ\x02o\x91Á§ó¦\x949\x8e®x=,©@u»;\x008À,À0\x00\x9fÌ©Ì¨ÌªÀ+À/\x00\x9eÀ$À(\x00kÀ#À'\x00gÀ")
<-- API down -->
Jan 08 05:17:56 <node3> patroni[3725]: 2019-01-08 05:17:56,092 DEBUG: API thread: <node1> - - [08/Jan/2019 05:17:56] code 400, message Bad HTTP/0.9 request type ('\x16\x03\x01\x02\x00\x01\x00\x01ü\x03\x03f')
Jan 08 05:17:56 <node3> patroni[3725]: 2019-01-08 05:17:56,093 DEBUG: API thread: <node3> - - [08/Jan/2019 05:17:56] code 400, message Bad request syntax ('\x16\x03\x01\x02\x00\x01\x00\x01ü\x03\x03')
Jan 08 05:17:56 <node3> patroni[3725]: 2019-01-08 05:17:56,095 DEBUG: API thread: <node1> - - [08/Jan/2019 05:17:56] code 400, message Bad HTTP/0.9 request type ('\x16\x03\x01\x00«\x01\x00\x00§\x03\x03Ú%í(Oa^ãEÑ#Y\x9bÖY\xadÞX·')
Jan 08 05:17:56 <node3> patroni[3725]: 2019-01-08 05:17:56,097 DEBUG: API thread: <node2> - - [08/Jan/2019 05:17:56] code 400, message Bad HTTP/0.9 request type ('\x16\x03\x01\x00«\x01\x00\x00§\x03\x03\x89®ÁTÝ\x00¸`?ëÉ\x8a×I\x91(cÈmM¬²^\x13')
Jan 08 05:17:56 <node3> patroni[3725]: 2019-01-08 05:17:56,098 DEBUG: API thread: <node3> - - [08/Jan/2019 05:17:56] code 400, message Bad HTTP/0.9 request type ('\x16\x03\x01\x02\x00\x01\x00\x01ü\x03\x03TDU\x9eù\x02µtÿYó\x08')
```
Then API was back again automatically.

**Comment by jouir** [comment]
When the API is down, failing host have an `ESTABLISHED` connection to API port:
```
node3> netstat -antp | grep <port> | grep ESTABLISHED
tcp        0      0 <node2>:<port>         <node3>:44164        ESTABLISHED 7077/python3
```

But, when we check on node3, nothing is available behind the client port:
```
node2> netstat -antp | grep -c 44164
0
```

**Comment by jouir** [comment]
We have released pressure of HAProxy probes by developing a little middleware to handle two basic routes (`/master` and `/replica`) and cache them for 1 second before contacting Patroni API. It's called [patroniglue](https://github.com/jouir/patroniglue). It may not be the cleanest code ever, that's why we called it "glue", but it does the job and load balancing is now safe.

**Comment by jouir** [comment]
We have updated to 1.6.0 including this [commit](https://github.com/zalando/patroni/commit/663026c34cc64f14040ceedf8a38f5094de41ba6#diff-790530624eaa21e5fb614fc321aef277R545-R547). No more API freeze.

**Comment by jouir** [comment]
We used to have this issue regularly on our backup hosts where we shut down Patroni every night to take a ZFS snapshot. They receive only writes from streaming replication. They will never become leader (`nofailover: true`), nor receive client traffic (`noloadbalance: true`) but can be used to bootstrap other nodes (`clonefrom: true`). In reality, bootstraps are quite rare. They were virtualized. As soon as we have upgraded them to baremetal servers and upgraded to 1.6.0 the issue was gone.

But, we still have virtualized clusters handling less traffic. And this night, we had an API freeze on the primary node of one of those clusters running Patroni version 1.6.0. That means #1039 doesn't solve this issue.

Monitoring system has detected the freeze at 18:26:02 EST with a resolution at 18:42:35 EST. Unfortunately, we had not enabled DEBUG mode because we thought this issue was solved so here are logs during that window:

```
Nov 20 18:26:03 node2 patroni[36968]: 2019-11-20 18:26:03,054 INFO: Lock owner: node2; I am node2
Nov 20 18:26:03 node2 patroni[36968]: 2019-11-20 18:26:03,059 INFO: no action.  i am the leader with the lock
Nov 20 18:26:13 node2 patroni[36968]: 2019-11-20 18:26:13,054 INFO: Lock owner: node2; I am node2
Nov 20 18:26:13 node2 patroni[36968]: 2019-11-20 18:26:13,057 INFO: no action.  i am the leader with the lock
...
Nov 20 18:40:43 node2 patroni[36968]: 2019-11-20 18:40:43,058 INFO: no action.  i am the leader with the lock
Nov 20 18:40:53 node2 patroni[36968]: 2019-11-20 18:40:53,054 INFO: Lock owner: node2; I am node2
Nov 20 18:40:53 node2 patroni[36968]: 2019-11-20 18:40:53,057 INFO: no action.  i am the leader with the lock
Nov 20 18:41:01 node2 systemd[1]: Reloading Runners to orchestrate a high-availability PostgreSQL.
Nov 20 18:41:01 node2 systemd[1]: Reloaded Runners to orchestrate a high-availability PostgreSQL.
Nov 20 18:41:03 node2 patroni[36968]: 2019-11-20 18:41:03,105 INFO: Changed DateStyle from ISO, MDY to iso, mdy
Nov 20 18:41:03 node2 patroni[36968]: 2019-11-20 18:41:03,106 INFO: PostgreSQL configuration items changed, reloading configuration.
Nov 20 18:41:03 node2 patroni[36968]: server signaled
Nov 20 18:41:03 node2 patroni[36968]: 2019-11-20 18:41:03,135 INFO: Lock owner: node2; I am node2
Nov 20 18:41:03 node2 patroni[36968]: 2019-11-20 18:41:03,140 INFO: no action.  i am the leader with the lock
Nov 20 18:41:13 node2 patroni[36968]: 2019-11-20 18:41:13,053 INFO: Lock owner: node2; I am node2
Nov 20 18:41:13 node2 patroni[36968]: 2019-11-20 18:41:13,058 INFO: no action.  i am the leader with the lock
Nov 20 18:41:23 node2 patroni[36968]: 2019-11-20 18:41:23,053 INFO: Lock owner: node2; I am node2
Nov 20 18:41:23 node2 patroni[36968]: 2019-11-20 18:41:23,056 INFO: no action.  i am the leader with the lock
Nov 20 18:41:28 node2 systemd[1]: Stopping Runners to orchestrate a high-availability PostgreSQL...
Nov 20 18:41:58 node2 systemd[1]: patroni.service: State 'stop-sigterm' timed out. Killing.
Nov 20 18:41:58 node2 systemd[1]: patroni.service: Killing process 36968 (patroni) with signal SIGKILL.
Nov 20 18:41:58 node2 systemd[1]: patroni.service: Main process exited, code=killed, status=9/KILL
Nov 20 18:41:58 node2 systemd[1]: Stopped Runners to orchestrate a high-availability PostgreSQL.
Nov 20 18:41:58 node2 systemd[1]: patroni.service: Unit entered failed state.
Nov 20 18:41:58 node2 systemd[1]: patroni.service: Failed with result 'timeout'.
Nov 20 18:41:58 node2 systemd[1]: Started Runners to orchestrate a high-availability PostgreSQL.
Nov 20 18:42:00 node2 patroni[21262]: 2019-11-20 18:42:00,167 INFO: Failed to import patroni.dcs.consul
Nov 20 18:42:00 node2 patroni[21262]: 2019-11-20 18:42:00,169 INFO: Failed to import patroni.dcs.etcd
Nov 20 18:42:00 node2 patroni[21262]: 2019-11-20 18:42:00,307 INFO: Failed to import patroni.dcs.kubernetes
Nov 20 18:42:00 node2 patroni[21262]: 2019-11-20 18:42:00,309 INFO: Connecting to <zk>:2181
Nov 20 18:42:00 node2 patroni[21262]: 2019-11-20 18:42:00,315 INFO: Zookeeper connection established, state: CONNECTED
Nov 20 18:42:00 node2 patroni[21262]: 2019-11-20 18:42:00,332 INFO: I am leader but not owner of the session. Removing leader node
Nov 20 18:42:00 node2 patroni[21262]: 2019-11-20 18:42:00,360 INFO: No PostgreSQL configuration items changed, nothing to reload.
Nov 20 18:42:00 node2 patroni[21262]: 2019-11-20 18:42:00,361 INFO: establishing a new patroni connection to the postgres cluster
Nov 20 18:42:00 node2 patroni[21262]: server signaled
Nov 20 18:42:00 node2 patroni[21262]: 2019-11-20 18:42:00,421 WARNING: Could not activate Linux watchdog device: "Can't open watchdog device: [Errno 13] Permission denied: '/dev/watchdog'"
Nov 20 18:42:00 node2 patroni[21262]: 2019-11-20 18:42:00,429 INFO: acquired session lock as a leader
Nov 20 18:42:00 node2 patroni[21262]: 2019-11-20 18:42:00,435 INFO: Lock owner: node2; I am node2
Nov 20 18:42:00 node2 patroni[21262]: 2019-11-20 18:42:00,446 INFO: no action.  i am the leader with the lock
Nov 20 18:42:00 node2 patroni[21262]: 2019-11-20 18:42:00,452 INFO: Lock owner: node2; I am node2
Nov 20 18:42:00 node2 patroni[21262]: 2019-11-20 18:42:00,453 INFO: no action.  i am the leader with the lock
```
Everything was fine from Patroni point of view, no action was taken. But API didn't respond. Load balancers did stop to send traffic because of that freeze. Logs show my restart to resume the API.

**Comment by jouir** [comment]
It's not limited to our virtualized environment. This week, we had the issue on our bare metal infrastructure on 4 nodes from 4 different clusters at the same time (in a 10 minutes window). 

On two nodes, we had multiple HTTP errors on Patroni logs:
```
Nov 26 20:37:58 node3 patroni[48023]: 2019-11-26 20:37:58,459 WARNING: Exception happened during processing of request from node4:40096
Nov 26 20:37:58 node3 patroni[48023]: 2019-11-26 20:37:58,459 WARNING: Traceback (most recent call last):
Nov 26 20:37:58 node3 patroni[48023]:   File "/usr/lib/python3.5/socketserver.py", line 625, in process_request_thread
Nov 26 20:37:58 node3 patroni[48023]:     self.finish_request(request, client_address)
Nov 26 20:37:58 node3 patroni[48023]:   File "/usr/lib/python3.5/socketserver.py", line 354, in finish_request
Nov 26 20:37:58 node3 patroni[48023]:     self.RequestHandlerClass(request, client_address, self)
Nov 26 20:37:58 node3 patroni[48023]:   File "/usr/lib/python3.5/socketserver.py", line 681, in __init__
Nov 26 20:37:58 node3 patroni[48023]:     self.handle()
Nov 26 20:37:58 node3 patroni[48023]:   File "/usr/lib/python3.5/http/server.py", line 422, in handle
Nov 26 20:37:58 node3 patroni[48023]:     self.handle_one_request()
Nov 26 20:37:58 node3 patroni[48023]:   File "/usr/lib/python3.5/http/server.py", line 400, in handle_one_request
Nov 26 20:37:58 node3 patroni[48023]:     if not self.parse_request():
Nov 26 20:37:58 node3 patroni[48023]:   File "/usr/lib/python3/dist-packages/patroni/api.py", line 414, in parse_request
Nov 26 20:37:58 node3 patroni[48023]:     ret = BaseHTTPRequestHandler.parse_request(self)
Nov 26 20:37:58 node3 patroni[48023]:   File "/usr/lib/python3.5/http/server.py", line 288, in parse_request
Nov 26 20:37:58 node3 patroni[48023]:     "Bad request version (%r)" % version)
Nov 26 20:37:58 node3 patroni[48023]:   File "/usr/lib/python3.5/http/server.py", line 477, in send_error
Nov 26 20:37:58 node3 patroni[48023]:     self.wfile.write(body)
Nov 26 20:37:58 node3 patroni[48023]:   File "/usr/lib/python3.5/socket.py", line 594, in write
Nov 26 20:37:58 node3 patroni[48023]:     return self._sock.send(b)
Nov 26 20:37:58 node3 patroni[48023]:   File "/usr/lib/python3.5/ssl.py", line 871, in send
Nov 26 20:37:58 node3 patroni[48023]:     return socket.send(self, data, flags)
Nov 26 20:37:58 node3 patroni[48023]: ConnectionResetError: [Errno 104] Connection reset by peer
```

We have a lot of those logs on the entire infrastructure. It may not be the root cause but a consequence.

On our ZooKeeper cluster (node anonymized to "zk1"), we have seen disconnections from nodes connecting via a network gateway (anonymized to "gw1"):

```
2019-11-26 20:38:01,719 - WARN  [NIOServerCxn.Factory:/zk1:2181:NIOServerCnxn@349] - caught end of stream exception
EndOfStreamException: Unable to read additional data from client sessionid 0x16c89ee2c0a0449, likely client has closed socket
        at org.apache.zookeeper.server.NIOServerCnxn.doIO(NIOServerCnxn.java:220)
        at org.apache.zookeeper.server.NIOServerCnxnFactory.run(NIOServerCnxnFactory.java:208)
        at java.lang.Thread.run(Thread.java:748)
2019-11-26 20:38:01,719 - INFO  [NIOServerCxn.Factory:/zk1:2181:NIOServerCnxn@1026] - Closed socket connection for client /gw1:48264 which had sessionid 0x16c89ee2c0a0449
2019-11-26 20:38:01,837 - INFO  [NIOServerCxn.Factory:/zk1:2181:NIOServerCnxnFactory@197] - Accepted socket connection from /gw1:56712
2019-11-26 20:38:01,837 - WARN  [NIOServerCxn.Factory:/zk1:2181:ZooKeeperServer@793] - Connection request from old client /gw1:56712; will be dropped if server is in r-o mode
2019-11-26 20:38:01,837 - INFO  [NIOServerCxn.Factory:/zk1:2181:ZooKeeperServer@832] - Client attempting to renew session 0x36c89eea6c104bf at /gw1:56712
2019-11-26 20:38:01,837 - INFO  [NIOServerCxn.Factory:/zk1:2181:Learner@107] - Revalidating client: 0x36c89eea6c104bf
2019-11-26 20:38:01,838 - INFO  [QuorumPeer[myid=1]/zk1:2181:ZooKeeperServer@595] - Established session 0x36c89eea6c104bf with negotiated timeout 4000 for client /gw1:56712
2019-11-26 20:38:05,927 - INFO  [NIOServerCxn.Factory:/zk1:2181:NIOServerCnxnFactory@197] - Accepted socket connection from /gw1:59036
2019-11-26 20:38:05,927 - WARN  [NIOServerCxn.Factory:/zk1:2181:ZooKeeperServer@793] - Connection request from old client /gw1:59036; will be dropped if server is in r-o mode
2019-11-26 20:38:05,927 - INFO  [NIOServerCxn.Factory:/zk1:2181:ZooKeeperServer@832] - Client attempting to renew session 0x36c89eea6c104bc at /gw1:59036
2019-11-26 20:38:05,927 - INFO  [NIOServerCxn.Factory:/zk1:2181:Learner@107] - Revalidating client: 0x36c89eea6c104bc
2019-11-26 20:38:05,928 - INFO  [QuorumPeer[myid=1]/zk1:2181:ZooKeeperServer@595] - Established session 0x36c89eea6c104bc with negotiated timeout 4000 for client /gw1:59036
2019-11-26 20:38:09,373 - INFO  [NIOServerCxn.Factory:/zk1:2181:NIOServerCnxnFactory@197] - Accepted socket connection from /gw1:56486
2019-11-26 20:38:09,373 - INFO  [NIOServerCxn.Factory:/zk1:2181:ZooKeeperServer@832] - Client attempting to renew session 0x26b52e70ec207f2 at /gw1:56486
2019-11-26 20:38:09,373 - INFO  [NIOServerCxn.Factory:/zk1:2181:Learner@107] - Revalidating client: 0x26b52e70ec207f2
2019-11-26 20:38:09,374 - INFO  [QuorumPeer[myid=1]/zk1:2181:ZooKeeperServer@595] - Established session 0x26b52e70ec207f2 with negotiated timeout 30000 for client /gw1:56486
2019-11-26 20:39:01,557 - INFO  [NIOServerCxn.Factory:/zk1:2181:NIOServerCnxnFactory@197] - Accepted socket connection from /zk1:37546
```
We thought it was network related. The netops team didn't notice unusual pattern. Other services were not impacted. Either they have automatically recovered from this flap, or there was no flap at all. Okay, let's deal with it.

Instead of rolling back to patroniglue, which is a pretty ugly software I wrote, we opt to install nginx as a reverse proxy to handle the HAProxy probes traffic. And we have changed HTTP method used by HAProxy to send GET requests instead of OPTIONS.

Example of HAProxy configuration:
```
listen read-write
    bind *:<read-write-port>
    option httpchk GET /read-write
    maxconn 1000000
    default-server fall 5 inter 1000 rise 5 downinter 1000 on-marked-down shutdown-sessions weight 10 maxconn 2000
    server node1 <ip1>:<postgres-port> check port <nginx-port> check-ssl verify none
    server node2 <ip2>:<postgres-port> check port <nginx-port> check-ssl verify none
    server node3 <ip3>:<postgres-port> check port <nginx-port> check-ssl verify none
```

Example of nginx configuration:
```
server {
  listen       *:<nginx-port> ssl;

  server_name  _;

  ssl on;
  ssl_certificate           /etc/ssl/certs/ssl-cert-snakeoil.pem;
  ssl_certificate_key       /etc/ssl/private/ssl-cert-snakeoil.key;
  ssl_session_cache         shared:SSL:10m;
  ssl_session_timeout       5m;
  ssl_protocols             TLSv1.2;
  ssl_ciphers               ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256:DHE-RSA-AES256-SHA;
  ssl_prefer_server_ciphers on;

  location ~ /\. {
    # Protection against versioning tools parsing
    return 404;
  }

  location ~ /(LICENSE|INSTALL|README|CHANGELOG|LASTGEN|MAINTAINERS|UPGRADE|AUTHORS|COPYRIGHT) {
    # Protection against auto generation
    return 404;
  }

  location / {
    limit_except GET {
        deny all;
    }
    proxy_pass            https://127.0.0.1:<patroni-api-port>/;
    proxy_read_timeout    10s;
    proxy_connect_timeout 30s;
    proxy_send_timeout    30s;
    proxy_set_header      Host $http_host;
    proxy_set_header      X-Real-IP $remote_addr;
    proxy_set_header      X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header      X-Forwarded-Proto https;
  }
}
```

**Comment by CyberDem0n** [comment]
Closing, it seems that https://github.com/zalando/patroni/pull/1547 should help.
